### ISSUE TITLE
multiple callback urls for applications

### ISSUE THREAD

**Issue by ccutrer** [isaacs/github#422]
we have a production version of our application, and then a "beta" version that is a snapshot of the production, and refreshed periodically. when using github to login to our app, it won't work with beta because we can only specify a single valid callback url. google, Facebook, linkedin, and others all support multiple callback urls to solve this situation nicely

**Comment by sgnn7** [comment]
+1. This is making it near impossible to make a generic gateway for hostname-routed application.

**Comment by ccutrer** [comment]
@sgnn7 for our application, we use a single hostname that all callbacks are routed to, and that hostname has special processing to extra information from the state parameter to determine where to redirect to to complete processing.

But that still doesn't work for our separate production/beta environments, because it will either send them to the production redirector service, which may not yet be running newer code to match any changes we made to the format of the state, or it will send them to the beta version of the redirector (what currently happens), and get an error about invalid callback url.

**Comment by sgnn7** [comment]
@ccutrer I edited my original post for clarity since I agree that it wasn't quite clear.

In our case we have multiple segments of our network that are separated by DNS zones (i.e. `http://*.foo.test.domain.com` and `http://*.bar.test.domain.com`) so for us and single-host matching we can't have fully isolated oauth'd endpoints without creating a ton of API keys or having that external host and it's doubly compounded since we use the hostname on top of the zone name (i.e http://somehost.bar.test.domain.com) on those segments to revproxy the info.

If we could use wildcards or domain suffixes, this could be solved easily but as-is it's a nightmare.

**Comment by ccutrer** [comment]
Yup. We have customer1.domain.com, customer2.domain.com, etc. That's production, and our app's callback URL is sso.domain.com which then forwards to the correct domain (after doing some validating; that host is the same servers as the main app, it just detects the special hostname and knows it should do the forwarding). Then we have customer1.test.domain.com, customer2.test.domain.com. (and staging and beta, and sometimes more). Huge pain.

**Comment by qrpike** [comment]
Agreed, makes developing a pain as well.

**Comment by wheattcom** [comment]
This makes testing and developing a total PITA, everybody else offers multiple callback urls (google, twitter, facebook, etc)

**Comment by stephen-walsh** [comment]
making things difficult without this feature

**Comment by qrpike** [comment]
Agreed!

**Comment by JamesXNelson** [comment]
Another drop in the bucket; a list of valid URLs seems like a reasonable solution.

Don't forget to email support@github.com instead of just +1ing here!

**Comment by BobVul** [comment]
The [`identity`](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/identity) API supported by both Firefox and Chrome extensions uses a different redirect domain on different browsers. Wildcards/suffixes won't help here. And having to host a server just to redirect auth responses for a client-side extension is fragile.

**Comment by JamesXNelson** [comment]
Hm.   Github already hosts an OAuth 2 server.  They just don't let you specify more than one callback url, which is ("should be") a fairly silly / trivial thing to fix.

**Comment by clepsydra** [comment]
I just found a Chrome extension that helps here too: **Requestly**.
Here you can define rules how Chrome shall redirect the user.
It's possible to make a rule which makes chrome open e.g. localhost:8080 when the redirect to the original redirect url happens.

**Comment by connor11528** [comment]
Yes how do you add localhost and production url for callback urls for Github apps?

**Comment by nathanchere** [comment]
Nice to see feedback on things like this is taken so seriously 🙄

**Comment by connor11528** [comment]
I got a response the recommended way for doing this is to set up a local and production application and use two sets of API keys, one for local and one for prod

**Comment by nathanchere** [comment]
That's great if you only have two environments, less great if you have many environments, totally hopeless if you have many environments with dynamically generated URLs (e.g. patch-142.test.myapp.com)

**Comment by nmatare** [comment]
Please do.

**Comment by SokratisVidros** [comment]
Any updates on this?

**Comment by Levi-Lesches** [comment]
@nathanchere, @SokratisVidros: Reminder that this is not [the official GitHub feedback repo](https://github.com/github/feedback/discussions/categories/general-feedback). Post there if you'd like an official response.
